File: cpp/28_实现strStr_easy.cpp
class Solution {
public:
    int strStr(string haystack, string needle) {
        if (needle.length() == 0) return 0;
        return KMP(haystack, needle);
    }
    
    int KMP(string &p, string &q) {
        int i = 0, j = -1;
        int m = p.length(), n = q.length();
        vector<int> next(n, -1);
        while (i < n-1) {
            if (j == -1 || q[i] == q[j]) {
                i++;
                j++;
                next[i] = j;
            }
            else
                j = next[j];
        }
        i = 0, j = 0;
        while (i < m && j < n) {
            if (j == -1 || p[i] == q[j]) {
                i++;
                j++;
            }
            else
                j = next[j];
        }
        if (j == n) return i-j;
        else return -1;
    }
};

File: python/28_实现strStr_easy.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        def KMP(p, q):
            next = [-1]
            m, n = len(p), len(q)
            i, j = 0, -1
            while (i < n-1):
                if j == -1 or q[i] == q[j]:
                    i += 1
                    j += 1
                    next.append(j)
                else:
                    j = next[j]
            i, j = 0, 0
            while (i<m and j<n):
                if j==-1 or p[i] == q[j]:
                    i += 1
                    j += 1
                else:
                    j = next[j]
            return i-j if j == n else -1
        if not len(needle):
            return 0
        return KMP(haystack, needle)
            